### PR TITLE
DEX-798 Fix api responses when incorrect public key

### PR DESCRIPTION
**Pull request by ngadiyak**

### CODE REVIEW

File: dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala [hunk 1]
```diff
@@ -588,6 +588,9 @@ case class SubscriptionsLimitReached(limit: Int, id: String)
       e"The limit of ${Symbol("limit") -> limit} subscriptions of this type was reached. The subscription of ${Symbol("id") -> id} was stopped"
     )
 
+case class InvalidPublicKey(reason: String)
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala:591]: We have `UserPublicKeyIsNotValid` which does same thing. But it doesn't contain the `reason`. Suggest:
1. Remove `InvalidPublicKey`
2. Add the `reason` to `UserPublicKeyIsNotValid`
3. Use it :)

**Review by vsuharnikov — CHANGES_REQUESTED** [review]

File: dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala [hunk 2]
```diff
@@ -180,6 +181,9 @@ class MatcherApiRoute(
       case Left(e) => complete(InfoNotFound(e))
     }
 
+  private def withValidPublicKey(addressOrError: Either[ValidationError.InvalidPublicKey, PublicKey])(f: PublicKey => Route): Route =
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala:184]: my bad

[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/http/routes/MatcherApiRoute.scala:184]: fixed

File: dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala (hunk 1 above)
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/balances/GetReservedBalanceSpec.scala [hunk 3]
```diff
@@ -56,9 +57,13 @@ class GetReservedBalanceSpec extends MatcherSuiteBase with TableDrivenPropertyCh
...
     shouldReturnErrorWithIncorrectApiKeyValue(dex1.rawApi.getReservedBalance(Base58.encode(alice.publicKey), incorrectApiKeyHeader))
 
-    //TODO: change after DEX-978
     "should return an error if publicKey is not correct base58 string" in {
-      validate404Exception(dex1.rawApi.getReservedBalance("null", System.currentTimeMillis, "sign"))
+      validateMatcherError(
+        dex1.rawApi.getReservedBalance("null", System.currentTimeMillis, "sign"),
+        StatusCodes.BadRequest,
+        12582912,
+        "Provided public key in not correct, reason: Invalid public key: Unable to decode base58: requirement failed: Wrong char 'l' in Base58 string 'null'"
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/balances/GetReservedBalanceSpec.scala:65]: > Provided public key in

**in** -> **is** ?

File: dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala [hunk 4]
```diff
@@ -535,7 +535,8 @@ case object ApiKeyIsNotProvided
 
 case object ApiKeyIsNotValid extends MatcherError(auth, commonEntity, commonClass, e"Provided API key is not correct")
 
-case object UserPublicKeyIsNotValid extends MatcherError(account, pubKey, broken, e"Provided user public key is not correct")
+case class UserPublicKeyIsNotValid(reason: String = "invalid public key")
+    extends MatcherError(account, pubKey, broken, e"Provided public key in not correct, reason: ${Symbol("reason") -> reason}")
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/error/MatcherError.scala:539]: here :)

**Comment by ngadiyak** [comment]
https://github.com/wavesplatform/matcher/pull/441

**Pull request closed without merging**